### ISSUE TITLE
LocalGeometry 'tracer_efit' not implemented for GENE

### ISSUE THREAD

**Issue by joeyschmidt97** [bpatel2107/pyrokinetics#255]
Hello,

I am trying to use pyrokinetics to read GENE files but am running into an error in that the geometry option 'tracer_efit' is not implemented for GENE code. Is there a way to suppress this error or a workaround? I really only need to read the output data (nrg, field, mom, etc.) files and not anything to do with the profiles.

Here's the code I am trying to run:
```
from pyrokinetics import Pyro

# Point to GENE input file
gene_file = 'filepath_to_simulation/scanfiles0000/parameters_0001'

# Load in file
pyro = Pyro(gk_file=gene_file, gk_code="GENE")

# Load in GENE output data
pyro.load_gk_output()
data = pyro.gk_output
```

Here's the error I am getting:
```
----> pyro = Pyro(gk_file=gene_file, gk_code="GENE")

File [~/.local/perlmutter/3.9-anaconda-2021.11/lib/python3.11/site-packages/pyrokinetics/pyro.py:867](https://vscode-remote+ssh-002dremote-002bperlmutter-002dp1-002enersc-002egov.vscode-resource.vscode-cdn.net/global/homes/fusion_research/~/.local/perlmutter/3.9-anaconda-2021.11/lib/python3.11/site-packages/pyrokinetics/pyro.py:867), in Pyro.read_gk_file(self, gk_file, gk_code, no_process)
    865 # Set LocalGeometry, LocalSpecies, Numerics, unless told not to.
    866 if "local_geometry" not in no_process:
--> 867     self.local_geometry = self.gk_input.get_local_geometry()
...
--> 189     raise NotImplementedError(
    190         f"LocalGeometry type {geometry_type} not implemented for GENE"
    191     )

NotImplementedError: LocalGeometry type tracer_efit not implemented for GENE
```

And in the geometry part of my GENE file. My initial thought is that the g-file (geomfile) is a relative path (loaded in wherever the GENE simulation directory is)

```
&geometry
magn_geometry = 'tracer_efit'
q0       =    7.1213618
shat     =    3.9870173
geomfile = 'g129015.00409_450'
edge_opt =    8.0000000
minor_r  =    1.0000000
major_R  =    1.0208728
rhostar  =   0.47368176E-02
dpdx_term= 'full_drift'
dpdx_pm  =   0.48974139E-01
norm_flux_projection  = F
/
```

Let me know if more information is required for this issue. Any help is appreciated!

**Comment by bpatel2107** [comment]
Hi and welcome to using pyrokinetics!

The reason for this is that pyrokinetics attempts to load in the `LocalGeometry` object (which stores $(R(\theta)$, $Z(\theta)$ and $B_{pol}(\theta)$ ) and currently we don't support the use of numerical equilibrium just yet.

The easiest thing to do is to just change the `magn_geometry = "miller"` as the only thing that is really needed from `LocalGeometry` when loading in the output is $q$ and $\hat{s}$ which are defined as normal. The only other concern here is that `trpeps` isn't defined, and in pyro will default to 0. This is only relevant when defining the value of $\partial \omega /\partial \rho$ used for the definition of $\gamma_{E\times B}$ and again doesn't technically matter for the output.

So artificially  setting `magn_geometry = "miller"` and `trpeps != 0.0 ` will likely let you read in the data.

The rigorous thing to do is to add in this ability to use numerical equilibria in `LocalGeometry` and figure out `trpeps` somehow.

Let me know if that works for you in the meantime and if not we can look into setting up numerical equilibria properly more quickly

**Comment by jfparisi** [comment]
Thanks Bhavin! I'm helping @joeyschmidt97 get set up w/ pyrokinetics. When reading in GENE files, I get a KeyError:
<img width="1725" alt="Screenshot 2023-09-21 at 1 00 08 PM" src="https://github.com/pyro-kinetics/pyrokinetics/assets/25325876/92e0c023-14a9-4529-b6c7-97a439e7309d">
MWE that produces this behavior:
```
import pyrokinetics as pk
import pathlib
gene_file='/p/gyro/jparisi/gene_stuff/scanfiles0003/parameters_0001'
pyro = pk.Pyro(gk_file=gene_file, gk_code = "GENE")
pyro.magn_geometry = "miller"
pyro.load_gk_output()
```
Any ideas for what's going wrong?

**Comment by bpatel2107** [comment]
The actual input file is stored in `pyro.gk_input.data` so you'd need to do

`pyro.gk_input.data["geometry"]["magn_geometry"] = "miller"`

That actual error seems to be from the fact that it can't find the `nrg` files. Are they stored in the same directory as the input file?

**Comment by joeyschmidt97** [comment]
I'm getting a different error `KeyError: 'simtimelim'` when I change the geometry to miller and put in a value for trpeps that is not 0 (is the numerical accuracy of this required or does it just have to not be zero?)

I changed the values in geometry to these values (all other values are the same):
```
&geometry
magn_geometry = 'miller'
trpeps = 0.83264625792299
...
/
```

I am pointing to a path that has all the files present and key this as the full error:
```
-> 1126 self.gk_output = read_gk_output(
   1127     path,
   1128     norm=local_norm,
   1129     load_fields=load_fields,
   1130     load_fluxes=load_fluxes,
   1131     load_moments=load_moments,
   1132     **kwargs,
   1133 )

File [~/.local/perlmutter/3.9-anaconda-2021.11/lib/python3.11/site-packages/pyrokinetics/gk_code/gk_output.py:833](https://vscode-remote+ssh-002dremote-002bperlmutter-002dp1-002enersc-002egov.vscode-resource.vscode-cdn.net/global/homes/j/fusion_research/~/.local/perlmutter/3.9-anaconda-2021.11/lib/python3.11/site-packages/pyrokinetics/gk_code/gk_output.py:833), in read_gk_output(path, norm, load_fields, load_fluxes, load_moments, gk_type, **kwargs)
    802 def read_gk_output(
    803     path: PathLike,
    804     norm: SimulationNormalisation,
   (...)
...
--> 180         return super(Namelist, self).__getitem__(lkey)
    181 else:
    182     keyiter = iter(key)

KeyError: 'simtimelim'
```

The code I am running is the same as in the first comment and I am running in a jupyter env.

Here's the full error:

**Comment by bpatel2107** [comment]
@joeyschmidt97 Looks likes its making progress and is actually trying to read the outputs. 

Sometimes in GENE you sent `['general']['simtimelim']` in the parameters file and if you reach that time GENE will spit out fluxes, moments etc., regardless of the time value so this is checked for. We don't handle cases where this isn't set (though would be easy to fix!).

Again for now in the `parameters_0001` please add `simtimelim=100000` or some large number in the `general` namelist. Once we get this working we can make a PR fixing all these issues at once!

**Comment by joeyschmidt97** [comment]
Got it to work and it's outputting data!

```
<pyrokinetics.GKOutput>
(Wraps <xarray.Dataset>)
Dimensions:         (theta: 512, kx: 1, ky: 1, time: 9304, field: 3,
                     species: 3, energy: 36, pitch: 8, flux: 3)
Coordinates:
  * time            (time) float64 0.0 0.000248 0.000496 ... 2.307 2.307 2.307
  * kx              (kx) float64 0.0
  * ky              (ky) float64 400.0
  * theta           (theta) float64 -9.425 -9.376 -9.327 ... 15.56 15.61 15.66
  * energy          (energy) float64 -1.0 -0.9429 -0.8857 ... 0.8857 0.9429 1.0
  * pitch           (pitch) float64 -1.0 -0.7143 -0.4286 ... 0.4286 0.7143 1.0
  * species         (species) <U8 'ion1' 'electron' 'ion2'
  * field           (field) <U4 'phi' 'apar' 'bpar'
  * flux            (flux) <U8 'particle' 'heat' 'momentum'
```

If the code could handle missing input cases that would be great to have! Mostly working with getting the output data.

Is there a shortcut in the mean time you recommend I do to by pass these missing input cases? My train of thought is to copy the parameters, modify the copied parameters to fill in missing data or swap variables (tracer_efit -> miller), and then use the copied parameters as file input for pyro. But let me know if there's a better way you recommend.

**Comment by bpatel2107** [comment]
That's probably the best way forward for now. The `simtimelim`/`trpeps` we could probably fix now but the numerical equilibrium will take some thinking about.

You can do it progranmatically by 

```python
from pyrokinetics import Pyro
from pyrokinetics.gk_code import GKInputGENE

gene_template = "/path/to/gene/parameters.dat"
gene_data_dir = gene_template.parent

gene = GKInputGENE()
gene.read_from_file(gene_template)

# Save original file and modify new one
gene.write(gene_data_dir / "original_parameters.dat")

if gene.data["geometry"]["magn_geometry"] != "miller":
    gene.data["geometry"]["magn_geometry"] = "miller"

if "simtimelim" not in gene.data["general"].keys():
    gene.data["general"]["simtimelim"] = 10000

if "trpeps" not in gene.data["geometry"].keys():
    gene.data["geometry"]["trpeps"] = 0.1

gene.write(gene_data_dir / gene_template)

pyro = Pyro(gk_file=gene_template)
pyro.load_gk_output()
```

**Comment by bpatel2107** [comment]
Fixed in #319
